rfcs/021-data_science_in_the_pipeline/README.md

```diff
@@ -0,0 +1,54 @@
+- Pipeline services do not know about data science
+- DS services do not know about SQS, message passing, the Actor model, etc
+- DS models are persistent and separate to DS inferrers
+- DS models can be retrained on demand
+
+These suggest 3 distinct, but loosely coupled, services:
+
+- **Inferrer**: A Python service that provides a synchronous API (most likely RESTful HTTP) that consumes whatever is needed to infer the new data, which it outputs.
+- **Model Trainer**: A Python service that can consume records from the catalogue index in bulk in order to train a model, and outputs/stores a persistent representation of this model for the *inferrer* to use.
+- **Inference Manager**: A Scala service that lives in the existing pipeline and contains "the usual" Wellcome message-passing, Akka, etc logic & libraries, which sychronously calls the `inferrer` and attaches the new data to the work/image (by populating a field) before passing it along.
+
+The usage of these services would look like this:
+
+![An architecture diagram for data science services](https://user-images.githubusercontent.com/4429247/76087593-631cc700-5fae-11ea-819f-3545e6dc7c65.png)
+
+#### Implementation details
+
+- The inferrer and inference manager exist in one task definition (and therefore on one host). The inferrer is not marked as essential: crashes are tolerated. Timeouts and other failures are naturally handled by the queue redrive policies.
+- The model trainer runs automatically as soon as it starts: it is triggered by an autoscaling policy. Scaling up could be based on regular time intervals, messages in queues, lambdas etc, and scaling down would be done by the trainer itself.
+- The inferrer loads the model from S3 when it starts. Inferrer instances will be short-lived (as they'll scale to zero when not in use) so triggering a restart or loading a new model by other means is not necessary.
+
+### Questions & drawbacks
+
+> Are the network requests for (eg) image files going to be very expensive and/or slow? How could we mitigate this?
+
+> How do we deal with wanting to infer several separate things for one piece of content (eg a feature vector _and_ a palette vector for a single image)? A chain of inferrer-manager pairs would be simplest and cleanest, but this would require `n` times as many network requests for images as one inferrer doing all the work.
+
+> Should DS APIs be open to the public?
+
+> Should model artefacts be open to the public?
+
+> Should DS inferrers work only for things we know (ie feature vectors can be inferred only for images within the WC ecosystem, by passing an image ID) or with any content (ie feature vectors can be inferred for any image, by passing a publicly accessible URI)
+
+> It seems like we'll have to reinfer on every existing work/image when a model is retrained, which will significantly add to the expense of retraining based on the above considerations about network requests etc. Is this necessarily the case?
```

Comment:
Aye. This is a good question - i'd like to understand how to handle this. Can we infer data about a new work from an existing model?

Reply:
We can and should infer data from a new work using an existing model, which is why we can confidently say that the model trainer will run infrequently. The problem identified here is that a new model would likely invalidate the existing inferred data in the case of (eg) vectors.

Reply:
But that's ok isn't it? The data in the catalogue API will be eventually consistent?

Reply:
Eventually but maybe with a period of being universally nonsensical in between, I think? @harrisonpim knows better

Reply:
We can definitely infer on new works with existing models - that's the whole point!

The (potential) issue is that we might add a bunch of new data to our catalogue which deviates significantly from what already exists, so would be poorly described by a model trained on the old data. In that case we'd like to retrain the model to describe the new data as well as the old stuff.

If the model aims to label things with eg plain text subjects (`cat`, `dog`, `house`), the model can be retrained to output new classes without invalidating the old inferred data - no issues there.

However, if the model outputs a _feature vector_ as its description of the image, text, or whatever, there's no guarantee that the space described by that feature vector has any correspondence with the one from the old model. Oversimplifying, it would be like one model learning to describe things in English, and a subsequent model learning to describe them in French. 
If we then want to compare descriptions of works (for image similarity, etc), we need them all to be described by the same model for that comparison to make sense. Unless we can figure out a way of updating models without invalidating the old descriptions, new descriptions would need to be inferred for every existing work, as well as the new ones.

Reply:
So feels like retraining necessitates a reindex (in some cases) - which we could certainly do, but is a motivation for ensuring that reindex cost is low. 

Another mitigation might be removing the "one message per work * number of services" overhead of a reindex and either:

- reindexing further along (i.e. after matching/merging)
- optionally have a way of batch updating inferred data